function/transform/special.go

```diff
@@ -157,3 +157,93 @@ var Alias = function.MetricFunction{
 		return list, nil
 	},
 }
+
+// Derivative is special because it needs to get one extra data point to the left
+// This transform estimates the "change per second" between the two samples (scaled consecutive difference)
+var Derivative = newDerivativeBasedTransform("derivative", derivative)
+
+func derivative(values []float64, parameters []function.Value, scale float64) ([]float64, error) {
+	result := make([]float64, len(values)-1)
+	for i := range values {
+		if i == 0 {
+			continue
+		}
+		// Scaled difference
+		result[i-1] = (values[i] - values[i-1]) / scale
+	}
+	return result, nil
+}
+
+// Rate is special because it needs to get one extra data point to the left
+// This transform functions exactly like Derivative but bounds the result to be positive.
+// That is, it returns consecutive differences which are at least 0.
+var Rate = newDerivativeBasedTransform("rate", rate)
+
+func rate(values []float64, parameters []function.Value, scale float64) ([]float64, error) {
+	result := make([]float64, len(values)-1)
+	for i := range values {
+		if i == 0 {
+			continue
+		}
+		// Scaled difference
+		result[i-1] = (values[i] - values[i-1]) / scale
+		if result[i-1] < 0 {
+			// values[i] is our best approximatation of the delta between i-1 and i
```

Comment:
+1 I would also love a context object on these functions so you could append notes a la EQ's processing logs. It would be nice to expose how many times the approximation was used so you could dig into lossy data.